src/yaml/next/serialization.cr

```diff
@@ -0,0 +1,203 @@
+module YAML
+  annotation Field
+  end
+
+  module Serializable
+    annotation Options
+    end
+
+    macro included
+      def self.new(ctx : YAML::ParseContext, node : YAML::Nodes::Node)
+        ctx.read_alias(node, \{{@type}}) do |obj|
+          return obj
+        end
+
+        instance = allocate
+
+        ctx.record_anchor(node, instance)
+
+        instance.initialize(ctx, node, nil)
```

Comment:
Can't you replace this hacky `.new` by removing the dummy argument then overriding the `.new` like this:
```cr
class Foo
  def initialize(@foo : String)
  end
  
  def self.new(foo : String)
    if rand > 0.5
      previous_def(foo)
    else
      Foo.new("bar")
    end
  end
end


p (1..10).map { Foo.new("foo") }
```

https://carc.in/#/r/41m4